File: sample-imagecapture/Sources/RTRDrawResultsView.m
/// ABBYY® Mobile Capture © 2019 ABBYY Production LLC.
/// ABBYY is a registered trademark or a trademark of ABBYY Software Ltd.

#import "RTRDrawResultsView.h"

#import <AbbyyRtrSDK/AbbyyRtrSDK.h>

@implementation RTRDrawResultsView {
	/// The background color for the rest of the screen outside the document boundary.
	UIColor* _areaFogColor;
}

- (id)initWithCoder:(NSCoder*)aDecoder
{
	self = [super initWithCoder:aDecoder];

	if(self != nil) {
		[self doInit];
	}
	
	return self;
}

- (id)initWithFrame:(CGRect)frame
{
	self = [super initWithFrame:frame];

	if(self != nil) {
		[self doInit];
	}

	return self;
}

- (void)doInit
{
	_areaFogColor = [[UIColor blackColor] colorWithAlphaComponent:0.7];
	self.exclusiveTouch = YES;
}

- (void)drawRect:(CGRect)rect
{
	[super drawRect:rect];

	CGContextRef currentContext = UIGraphicsGetCurrentContext();

	CGContextSaveGState(currentContext);
	CGContextTranslateCTM(currentContext, 0, 0);
	CGContextScaleCTM(currentContext, CGRectGetWidth(self.bounds) / self.imageBufferSize.width,
		CGRectGetHeight(self.bounds) / self.imageBufferSize.height);

	[self drawBlocksInContext:currentContext];
	[self drawFogOverDocumentInContext:currentContext];

	CGContextRestoreGState(currentContext);
}

- (void)drawBlocksInContext:(CGContextRef)context
{
	CGContextSaveGState(context);
	CGContextRef currentContext = UIGraphicsGetCurrentContext();

	if(self.documentBoundary.count != 0) {
		[self addPathForDocumentBoundaryToContext:currentContext];
		CGContextClip(currentContext);
	}

	for(RTRQualityAssessmentForOCRBlock* block in self.blocks) {
		UIColor* color = nil;
		BOOL fillRect = NO;
		switch(block.type) {
			case RTRQualityAssessmentForOCRTextBlock:
				color = [UIColor colorWithRed:((100.f - block.quality) / 100.f) green:(block.quality / 100.f) blue:0 alpha:0.4f];
				fillRect = YES;
				break;
			case RTRQualityAssessmentForOCRUnknownBlock:
				color = [[UIColor lightGrayColor] colorWithAlphaComponent:0.2f];
				break;
			default:
				break;
		}
		if(color != nil) {
			[self drawImageQuadrangle:block.rect color:color fill:fillRect context:currentContext];
		}
	}

	CGContextRestoreGState(context);
}

- (void)addPathForDocumentBoundaryToContext:(CGContextRef)context
{
	NSArray* boundary = self.documentBoundary;
	CGPoint firstPoint = [boundary.firstObject CGPointValue];
	CGContextMoveToPoint(context, firstPoint.x, firstPoint.y);
	for(NSValue* obj in boundary) {
		CGContextAddLineToPoint(context, obj.CGPointValue.x, obj.CGPointValue.y);
	}

	CGContextClosePath(context);
}

- (void)drawImageQuadrangle:(CGRect)blockRect color:(UIColor*)color fill:(BOOL)fill context:(CGContextRef)context
{
	CGContextSetStrokeColorWithColor(context, color.CGColor);
	CGContextStrokeRect(context, blockRect);
	if(fill) {
		CGContextSetFillColorWithColor(context, color.CGColor);
		CGContextFillRect(context, blockRect);
	}
}

- (void)drawFogOverDocumentInContext:(CGContextRef)context
{
	if(self.documentBoundary.count == 0) {
		return;
	}
	CGContextSaveGState(context);

	CGRect scaledBounds = CGRectMake(0, 0, self.imageBufferSize.width, self.imageBufferSize.height);

	CGContextAddRect(context, scaledBounds);
	[self addPathForDocumentBoundaryToContext:context];
	CGContextEOClip(context);

	// Fill the background outside the area of interest
	CGContextSetFillColorWithColor(context, _areaFogColor.CGColor);
	CGContextFillRect(context, scaledBounds);

	CGContextRestoreGState(context);
}

- (void)clear
{
	self.blocks = nil;
	self.documentBoundary = nil;

	[self setNeedsDisplay];
}

@end

File: sample-imagecapture-swift/Sources/RTRDrawResultsView.swift
/// ABBYY® Mobile Capture © 2019 ABBYY Production LLC.
/// ABBYY is a registered trademark or a trademark of ABBYY Software Ltd.

import AbbyyRtrSDK

class RTRDrawResultsView: UIView {
	/// The background color for the rest of the screen outside the document boundary.
	private let areaFogColor: UIColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0.7)

	var documentBoundaryPoints : Array<CGPoint>? {
		get {
			guard let boundary = documentBoundary else {
				return nil
			}

			var result: Array<CGPoint> = []
			for pointValue in boundary {
				result.append(pointValue.cgPointValue)
			}
			return result
		}
	}

	/// Image size, to scale coordinates.
	var imageBufferSize: CGSize = CGSize.zero
	/// Found document boundary.
	var documentBoundary: Array<NSValue>? = nil
	/// Quality assessment blocks.
	var blocks: Array<RTRQualityAssessmentForOCRBlock>? = nil

// MARK: - LifeCycle

	override init(frame: CGRect)
	{
		super.init(frame: frame)
		self.doInit()
	}

	required init?(coder aDecoder: NSCoder)
	{
		super.init(coder: aDecoder)
		self.doInit()
	}

	/// Clear view.
	func clear()
	{
		documentBoundary = nil
		blocks = nil
		setNeedsDisplay()
	}

// MARK: - Private

	private func doInit()
	{
		isExclusiveTouch = true
	}

	override func draw(_ rect: CGRect)
	{
		super.draw(rect)

		if let currentContext = UIGraphicsGetCurrentContext() {
			currentContext.saveGState()
			currentContext.translateBy(x: 0, y: 0)
			currentContext.scaleBy(x: bounds.width / imageBufferSize.width, y: bounds.height / imageBufferSize.height)

			drawBlocks(context: currentContext)
			drawFogOverDocumentBoundary(context: currentContext)

			currentContext.restoreGState()
		}
	}

	private func drawBlocks(context: CGContext)
	{
		context.saveGState()

		if let currentContext: CGContext = UIGraphicsGetCurrentContext() {
			if let boundary = documentBoundary {
				if boundary.count != 0 {
					addPathForDocumentBoundary(context: currentContext)
					currentContext.clip()
				}
			}

			if let blocksToDraw = blocks {
				for block in blocksToDraw {
					var color: UIColor?
					var fillRect = false
					switch(block.type) {
						case .textBlock:
							color = UIColor.init(red: (1 - CGFloat(block.quality) / 100), green: CGFloat(block.quality) / 100, blue: 0, alpha: 0.4)
							fillRect = true
						case .unknownBlock:
							color = UIColor.lightGray.withAlphaComponent(0.2)
					}
					if let blockColor = color {
						drawBlock(rect: block.rect, color: blockColor, fill: fillRect, context: currentContext)
					}
				}
			}
		}
		context.restoreGState()
	}

	private func drawBlock(rect: CGRect, color: UIColor, fill: Bool, context: CGContext)
	{
		context.setStrokeColor(color.cgColor)
		context.stroke(rect)
		if fill {
			context.setFillColor(color.cgColor)
			context.fill(rect)
		}
	}

	private func drawFogOverDocumentBoundary(context: CGContext)
	{
		if let boundary = documentBoundary {
			if boundary.count == 0 {
				return
			}

			context.saveGState()
			let scaledBounds = CGRect.init(origin: CGPoint.zero, size: imageBufferSize)
			context.addRect(scaledBounds)
			addPathForDocumentBoundary(context: context)
			context.clip(using: CGPathFillRule.evenOdd)

			context.setFillColor(areaFogColor.cgColor)
			context.fill(scaledBounds)

			context.restoreGState()
		}
	}

	private func addPathForDocumentBoundary(context: CGContext)
	{
		addPathForPoints(context: context, points: documentBoundaryPoints)
	}

	private func addPathForPoints(context: CGContext!, points: Array<CGPoint>?)
	{
		if let _points = points {
			context.addLines(between: _points)
			context.closePath()
		}
	}
}
